### PR TITLE
feat/ #1180 refactor gen_begin_tx_ops and gen_end_tx_ops

### LINKED ISSUE #1180: Move into struct-associated functions EndTx and BeginTx `ExecutionState`s
See: 
- https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/bus-mapping/src/circuit_input_builder.rs#L256-L261
- https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/bus-mapping/src/circuit_input_builder.rs#L274-L279

### PR DESCRIPTION
**Pull request by KimiWu123**
### Description

fix #1180 

### Issue Link

#1180 

### Type of change

- [x] New feature (non-breaking change which adds functionality)

### Contents

see #1180 

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?

Pass all the existing test cases

### CODE REVIEW

File: bus-mapping/src/circuit_input_builder.rs [hunk 1]
```diff
@@ -270,11 +270,11 @@ impl<'a> CircuitInputBuilder {
         let mut tx = self.new_tx(eth_tx, !geth_trace.failed)?;
         let mut tx_ctx = TransactionContext::new(eth_tx, geth_trace, is_last_tx)?;
 
-        // TODO: Move into gen_associated_steps with
-        // - execution_state: BeginTx
-        // - op: None
```
Comment (KimiWu123) [inline bus-mapping/src/circuit_input_builder.rs:275]: Add `op` as arugment when we need it. It's not necessary for now.

File: bus-mapping/src/evm/opcodes/begin_end_tx.rs [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+            ExecState::EndTx => gen_end_tx_steps(state),
+            _ => {
+                evm_unimplemented!("unsupported tx execution step {:?}", execution_step);
+                Ok(state.new_begin_tx_step())
+            }
+        }
+    }
+}
+
+fn gen_begin_tx_steps(state: &mut CircuitInputStateRef) -> Result<ExecStep, Error> {
```
Comment (KimiWu123) [inline bus-mapping/src/evm/opcodes/begin_end_tx.rs:33]: just moved from `bus-mapping/src/evm/opcodes.rs` directly, no any code changes.

File: bus-mapping/src/evm/opcodes/begin_end_tx.rs [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+            ] {
+                state.call_context_write(&mut exec_step, call.call_id, field, value);
+            }
+
+            Ok(exec_step)
+        }
+    }
+}
+
+fn gen_end_tx_steps(state: &mut CircuitInputStateRef) -> Result<ExecStep, Error> {
```
Comment (KimiWu123) [inline bus-mapping/src/evm/opcodes/begin_end_tx.rs:229]: just moved from `bus-mapping/src/evm/opcodes.rs` directly, no any code changes.

File: bus-mapping/src/evm/opcodes.rs [hunk 4]
```diff
@@ -350,307 +363,19 @@ pub fn gen_associated_ops(
...
-            Ok(exec_step)
+pub fn gen_associated_steps(
+    state: &mut CircuitInputStateRef,
+    execution_step: ExecState,
+) -> Result<ExecStep, Error> {
+    let fn_gen_associated_steps = match execution_step {
+        ExecState::BeginTx | ExecState::EndTx => BeginEndTx::gen_associated_steps,
+        _ => {
+            evm_unimplemented!("unsupported execution step {:?}", execution_step);
+            Dummy::gen_associated_steps
```
Comment (han0110) [inline bus-mapping/src/evm/opcodes.rs:374]: I think we could just mark it as `unreachable` here:

```suggestion
            unreachable!()
```

File: bus-mapping/src/evm/opcodes/begin_end_tx.rs [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+    fn gen_associated_steps(
+        state: &mut CircuitInputStateRef,
+        execution_step: ExecState,
+    ) -> Result<ExecStep, Error> {
+        match execution_step {
+            ExecState::BeginTx => gen_begin_tx_steps(state),
+            ExecState::EndTx => gen_end_tx_steps(state),
+            _ => {
+                evm_unimplemented!("unsupported tx execution step {:?}", execution_step);
+                Ok(state.new_begin_tx_step())
```
Comment (han0110) [inline bus-mapping/src/evm/opcodes/begin_end_tx.rs:27]: Same as above

**Review by han0110 — APPROVED** [review]
LGTM!

File: bus-mapping/src/evm/opcodes/begin_end_tx.rs (hunk 5 above)
Comment (KimiWu123) [reply bus-mapping/src/evm/opcodes/begin_end_tx.rs]: fixed in fixed in [b737885](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1336/commits/b737885318cb7fc714406a0735502980d40db5cb)

**Comment by ed255** [comment]
@icemelon could you help us assign some review from scroll for this PR? Thanks!

**Comment by lispc** [comment]
why combining begin_tx/end_tx together? in evm circuit side, they are seperated.  I think it is not necessary to put them together.   Separating them as begin_tx::BeginTx and end_tx::EndTx (into 2 rust files) seesm more consistent with evm circuit side.

File: bus-mapping/src/evm/opcodes.rs [hunk 6]
```diff
@@ -133,6 +136,16 @@ impl Opcode for Dummy {
     }
 }
 
+#[allow(unused_variables)]
+impl TxExecSteps for Dummy {
+    fn gen_associated_steps(
+        state: &mut CircuitInputStateRef,
+        execution_step: ExecState,
+    ) -> Result<ExecStep, Error> {
+        Ok(state.new_begin_tx_step())
```
Comment (lispc) [inline bus-mapping/src/evm/opcodes.rs:145]: i think either `state.new_step(&geth_steps[0])` or even `unimplemented!("...")` would be better implementation for this. `state.new_step(&geth_steps[0])` seems to be good default implementation, while `unimplemented` at least it reports error. While, state.new_begin_tx_step() is a wrong implementation for most cases.

Comment (KimiWu123) [reply bus-mapping/src/evm/opcodes.rs]: just realized `impl TxExecSteps for Dummy` could be reomved and fixed in [1dcc3ac](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1336/commits/1dcc3ac0bb483108471a70d2c054cdcdbff955ec)

**Comment by KimiWu123** [comment]
> why combining begin_tx/end_tx together? in evm circuit side, they are seperated. I think it is not necessary to put them together. Separating them as begin_tx::BeginTx and end_tx::EndTx (into 2 rust files) seesm more consistent with evm circuit side.
> 
> @han0110 do you think the trait `TxExecSteps` useful? Calling gen_begin_tx_steps/gen_end_tx_steps is easier to understand
> 
> i would still suggest split begin tx and end tx into 2 files

Actually, I felt similar while refactoring this. Could @CPerezz give us more idea about this refactor behind?

**Comment by han0110** [comment]
I also feel `TxExecSteps` seems not so useful because I couldn't think about extra variant for it, just splitting it them into 2 files seems more straightforward to me.

**Comment by ed255** [comment]
@CPerezz could you please help reviewing this?

**Review by CPerezz — COMMENTED** [review]
I  don't think this was he purpose of the issue.

The main idea was that `end_tx` `begin_tx` are tho functiopns that share a lot of code. Therefore, it would be nice to reduce the code duplications.

Hence, the main idea is that in the trait added here, most of this shared code is there. Instead of beiing in the functions anyways.

**Pull request merged**